sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -2466,3 +2466,37 @@ object RemoveRepetitionFromGroupExpressions extends Rule[LogicalPlan] {
       }
   }
 }
+
+object EliminateCast extends Rule[LogicalPlan] {
```

Comment:
I think you should improve `UnwrapCastInBinaryComparison`.

Reply:
ok, thanks. I will make a try

Reply:
@HyukjinKwon Hi, I checked UnwrapCastInBinaryComparison, and this rule may be applied on all filter conditions, including non-partitioned field, which may cause some bug if modify this rule. Maybe, modifying the `convertFilters` method in `HiveShim.scala` is a better way.